src/main/java/net/neoforged/neoforge/network/registration/ChannelAttributes.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.network.registration;
+
+import io.netty.channel.Channel;
+import io.netty.util.AttributeKey;
+import java.util.EnumMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import net.minecraft.network.Connection;
+import net.minecraft.network.ConnectionProtocol;
+import net.minecraft.resources.ResourceLocation;
+import net.neoforged.neoforge.network.connection.ConnectionType;
+import net.neoforged.neoforge.network.payload.CommonRegisterPayload;
+import net.neoforged.neoforge.network.payload.MinecraftRegisterPayload;
+import net.neoforged.neoforge.network.payload.ModdedNetworkPayload;
+import net.neoforged.neoforge.network.payload.ModdedNetworkQueryPayload;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Utilities for manipulation of Netty {@link Channel} attributes
+ */
+@ApiStatus.Internal
+public class ChannelAttributes {
+    /**
+     * Negotiated modded payload setup from {@link ModdedNetworkQueryPayload} and/or {@link ModdedNetworkPayload}.
+     */
+    public static final AttributeKey<NetworkPayloadSetup> PAYLOAD_SETUP = AttributeKey.valueOf("neoforge:payload_setup");
+
+    /**
+     * Known ad-hoc channels from {@link MinecraftRegisterPayload}.
+     */
+    public static final AttributeKey<Set<ResourceLocation>> ADHOC_CHANNELS = AttributeKey.valueOf("neoforge:adhoc_channels");
+
+    /**
+     * Known common channels from {@link CommonRegisterPayload}.
+     */
+    public static final AttributeKey<Map<ConnectionProtocol, Set<ResourceLocation>>> COMMON_CHANNELS = AttributeKey.valueOf("neoforge:common_channels");
```

Comment:
Why not a `SetMultimap` from Guava?

Reply:
Don't know, but my source attachment is busted for guava in neodev atm, so I shall be leaving it as-is.

Reply:
Perhaps then, we should add a comment to consider moving it to SetMultimap in the future?

Reply:
Eh, probably not necessary.  If someone wants to move it they can do so with or without the todo.